Updates to link styles and link hover states

frankieroberto opened the pull request:
This updates links (and link-like elements such as details) so that the underline is consistently thinner and offset further away from the link text, improving readability.

Links also have a clearer hover state using dark blue (`#003087`) instead of dark pink (`#7c2855`), and the underline gets thicker instead of disappearing, to make the link stand out more to users.

It also introduces new modifier classes for:

- text colour links (`nhsuk-link--text-colour`)
- links without underlines (`nhsuk-link--no-underline`)

<br>

| Before | After |
|----|-----|
| <img width="636" height="912" alt="Links before" src="https://github.com/user-attachments/assets/dbf479e1-6554-461d-8302-dad8b7f8f9a8" />  | <img width="636" height="912" alt="Links after" src="https://github.com/user-attachments/assets/5caeb8e1-e932-43ec-83a0-a45ae1a3cacd" /> |

## Checklist

- [ ] Decide which components should have underlines and which should not
- [ ] Tested against our [testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [x] Follows our [coding standards and style guide](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/coding-standards.md)
- [x] CHANGELOG entry

paulrobertlloyd commented:
Should the active colour be different to the hover colour, maybe use the black colour variable (I think it’s black on GOV.UK)?

frankieroberto commented:
> Should the active colour be different to the hover colour, maybe use the black colour variable (I think it’s black on GOV.UK)?

I had the same thought, wasn't sure how necessary it was to visually differentiate hover from active for links.

I’ve changed the active colour to black in [5a70da7](https://github.com/nhsuk/nhsuk-frontend/pull/1367/commits/5a70da76d9b323f942e116718a3cd90a0e5e93df), which matches GOVUK. It's pretty subtle though!

Also not sure if `#002f5c` (`nhsuk-shade($nhsuk-link-color, 50%);`) is the best shade of dark blue to use or not. It looks slightly muddy to my eyes but maybe it's fine?

edwardhorsford commented:
Can you show some screenshots / a video? Do we know why the NHS design system originally used different colours?

frankieroberto commented:
@edwardhorsford @anandamaryon1 @paulrobertlloyd @colinrotherham review apps now enabled, so you can preview it here: https://nhsuk-frontend-pr-1367.herokuapp.com/nhsuk-frontend/ (scroll the bottom for a page with all the link examples) 🎉

edwardhorsford commented:
Should this be checked with the NHS.UK and NHS App teams?

colinrotherham commented:
> Should this be checked with the NHS.UK and NHS App teams?

Yeah should think so. I'm just bringing https://github.com/nhsuk/nhsuk-frontend/pull/1366 up to date for visibility

Hope that's alright

frankieroberto commented:
@anandamaryon1 they're technically separate, so could be merged in separately, but might make sense to release them together?

frankieroberto commented:
@anandamaryon1 @colinrotherham @paulrobertlloyd @edwardhorsford I’ve updated this PR so it it’s up-to-date with main, and now also contains the changes in #366 to add the underline offset and underlining thickening on hover.

Happy to take suggestions on further tweaks. @colinrotherham suggested changing the underline offset from 3px to 2.5px on the other PR.

I’ve also added a regular underline to the action links, which I think was by accident, but with the underline offset I wonder if it’s actually an improvement?

We could separately make a decision on whether to keep the underlines in the header or not.

Review app here: https://nhsuk-frontend-pr-1367.herokuapp.com/nhsuk-frontend/

anandamaryon1 commented:
> What about a super minor change to ~3px~ 2.5px?

I did try this in-browser and I don't find any difference? It snaps to either 2px or 3px on my low DPI screen, and on the high-DPI screen it's not really an issue anyway.

colinrotherham commented:
>I did try this in-browser and I don't find any difference? It snaps to either 2px or 3px on my low DPI screen, and on the high-DPI screen it's not really an issue anyway.

Ah shame, so you're still seeing what you posted in https://github.com/nhsuk/nhsuk-frontend/pull/1366#issuecomment-3297708885 instead of the (subtly) better "g" above?

```scss
$nhsuk-link-underline-thickness: string.unquote("max(1px, #{math.div(1px, 16px) * 1rem})") !default;
$nhsuk-link-underline-offset: math.div(2.5px, 19px) * 1em !default;
```

☝️ If anyone else wants to try, here are the variables expressed with units not magic numbers

paulrobertlloyd commented:
Worth pointing out that the link offset also needs to be applied to the account button in the header: 

<img width="380" height="65" alt="Screenshot of button in header account area." src="https://github.com/user-attachments/assets/05c84fa9-2ae3-41cf-8f7f-bf0b0d20a10a" />

I think I’d recommend not linking anything in the header, except for links inside the account area. There’s a sorta precedent given a linked service name is not underlined either.

anandamaryon1 commented:
We reviewed the new link styles and colours with the design system review panel recently and in general they're happy with them and ok to proceed. 

The concerns are around specific components and whether we should underline all links or be more selective, eg. in the header nav items or action link, where there is already a good expectation that they're clickable. 

I'm going to do some mockups of a header and other components without underlines and consider some different hover states. 

If we go ahead with removing underlines from some components, it'll be quite a significant change so we'll need to do some testing, which may take some time, so perhaps we need an interim solution too. Which is probably to implement these underline styles and not change what is and isn't underlined at this time.

frankieroberto commented:
@colinrotherham @anandamaryon1 @paulrobertlloyd I've updated this to fix the conflicts with `main`, and also to remove the underline from the action link.

This should now mean that this PR doesn’t make any changes to which components are underlined by default - ie header links and tabs remain underlined, action links remain non-underlined.

So the only differences now are:

* underlines have an offset
* on hover, instead of the underlines disappearing, the underline remains but gets thicker
* the hover link colour changes from purple to dark blue (except where reversed or error summaries where it stays the same colour)

I'm not sure whether we count this as a breaking change or not, but we could still hold it back for v11 as it's a significant visual change.

If we decide to change which components should be underlined (eg removing it from header nav links, adding it to action links) those could be follow-up PRs (eg #1641)?

@paulrobertlloyd I think both the visual bugs you spotted above (header log out button and expander focus style) are now fixed.

anandamaryon1 commented:
> @anandamaryon1 I’m sure you’re aware of these, but sharing screenshots of focused details and expander links from your test environment that has these new link styles:
> 

@paulrobertlloyd 
Oh don't look at the focus states! 🙈 I was just playing with hover styles, not looked at other states at all on those tests. 

@frankieroberto 
Great, that's helpful thanks, agree we'll want any significant changes to what gets underlined as a separate PR. Yes maybe holding for v11 makes sense, it's a subtle but very far reaching visual change.

frankieroberto commented:
@colinrotherham looks good to me, thanks for the rebase!

colinrotherham commented:
@frankieroberto I've had a go at adding a changelog entry so we can get this merged

frankieroberto commented on CHANGELOG.md:
```diff
@@ -2,6 +2,63 @@
+
+- `nhsuk-link-style-no-underline` mixin in your Sass, or
+- `nhsuk-link--no-underline` class in your HTML
+
```
It took me a minute to spot that one of these was a mixin and the other was an HTML class.

Perhaps clearer to separate them a bit more and add `@include` for the mixin?

```suggestion
You can now remove underlines from links by adding the `nhsuk-link--no-underline` HTML class.

If you are creating a custom component you can use the Sass mixin:

```scss
@include nhsuk-link-style-no-underline;
```

frankieroberto commented on packages/nhsuk-frontend/src/nhsuk/core/elements/_links.scss:
```diff
@@ -26,16 +26,37 @@ a {
 
-.nhsuk-link--no-visited-state {
-  @include nhsuk-link-style-no-visited-state;
+// Variant classes should always be used in conjunction with the .nhsuk-link
```
Currently the `.nhsuk-link` class doesn’t actually do anything at all, and we rely on the `a { }` selector above?

Should we either change it to `a, .nhsuk-link {}` - or leave it and update this comment?

colinrotherham replied:
Updated in https://github.com/nhsuk/nhsuk-frontend/pull/1367/commits/7014be0ff7e088f9ea83a781752b16ce005d2fee

frankieroberto commented on packages/nhsuk-frontend/src/nhsuk/core/elements/_links.scss:
```diff
@@ -26,16 +26,37 @@ a {
+
+.nhsuk-link--muted {
+  @include nhsuk-link-style-muted;
+}
```
@colinrotherham @anandamaryon1 did I add 'muted' or did that come later? I honestly can't remember!

Did we have an example for where this’d be used? And was there a reason not to use 'secondary', like we do for the override class?

colinrotherham replied:
It was me in https://github.com/nhsuk/nhsuk-frontend/pull/1366 via this commit 😬

* [`5115b84` Add link modifiers from GOV.UK Frontend](https://github.com/nhsuk/nhsuk-frontend/pull/1366/commits/5115b841fbbe67daada2174ed947662bc91525ee)

Yeah it makes sense to use 'secondary' but only if we want to keep it though?

* `.nhsuk-link--text-colour`
* `.nhsuk-link--secondary-text-colour`

It's not documented but was previously used for feedback links ([see comment](https://github.com/alphagov/govuk-design-system/issues/1603#issuecomment-822477055))

<img width="563" height="166" alt="Muted feedback link highlighted in a red outline" src="https://github.com/user-attachments/assets/75a02c38-5653-4cba-96a6-3a40697fce64" />

<br>
<br>

It's usage is very limited so I'd be happy to remove it:

<img width="672" height="346" alt="Muted 'view all' link highlighted in a red outline" src="https://github.com/user-attachments/assets/a4a52f79-f1ad-4286-8d97-1b3fe90614da" />

anandamaryon1 commented:
Pushed a fix, needed to add text-decoration underline back on when hovering, after it is removed from unhovered service names.

frankieroberto commented:
> Pushed a fix, needed to add text-decoration underline back on when hovering, after it is removed from unhovered service names.

@anandamaryon1 I was working on the same but you beat me to the commit! 😄 I’ve updated the reference image and removed a trailing space though.
